File: app/src/main/java/com/example/vivic/nolost/Login/LoginActivity.java
package com.example.vivic.nolost.Login;


import android.arch.lifecycle.ViewModelProviders;
import android.content.Context;
import android.os.Bundle;
import android.support.v7.widget.AppCompatEditText;
import android.text.TextUtils;
import android.util.Log;
import android.view.View;
import android.view.inputmethod.EditorInfo;
import android.view.inputmethod.InputMethodManager;
import android.widget.Button;
import android.widget.ImageView;

import com.example.vivic.nolost.GlideApp;
import com.example.vivic.nolost.R;
import com.example.vivic.nolost.activity.BaseActivity;
import com.example.vivic.nolost.bean.MyUser;
import com.example.vivic.nolost.commonUtil.BindEventBus;
import com.example.vivic.nolost.commonUtil.NetworkUtil;
import com.example.vivic.nolost.commonUtil.NoDoubleClickListener;
import com.example.vivic.nolost.commonUtil.pref.CommonPref;
import com.example.vivic.nolost.commonUtil.toastUtil.ToastUtil;
import com.xiasuhuei321.loadingdialog.view.LoadingDialog;

import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;
import org.greenrobot.eventbus.ThreadMode;

import cn.bmob.v3.exception.BmobException;
import cn.sharesdk.sina.weibo.SinaWeibo;

@BindEventBus
public class LoginActivity extends BaseActivity {
    private static final String TAG = "LoginActivity";

    private LoginViewModel loginViewModel;
    private ImageView ivIcon;
    private String userAccount;
    private String userPassword;
    private AppCompatEditText etAccount;
    private AppCompatEditText etPassword;
    private InputMethodManager inputMethodManager;
    private ILoginCallback<MyUser> iSdkLoginCallback;
    private LoadingDialog loadingDialog;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_login);
        loginViewModel = ViewModelProviders.of(this).get(LoginViewModel.class);
        inputMethodManager = (InputMethodManager) getSystemService(Context.INPUT_METHOD_SERVICE);
        initview();
        thirdLogin();
        initCallback();
    }


    private void initview() {
        Button btnLogin = findViewById(R.id.btn_login);
        Button btnSign = findViewById(R.id.btn_sign);
        etAccount = findViewById(R.id.et_user_account);
        etPassword = findViewById(R.id.et_password);
        etPassword.setOnEditorActionListener((v, actionId, event) -> {
            if (actionId == EditorInfo.IME_ACTION_DONE) {
                inputMethodManager.hideSoftInputFromWindow(etPassword.getWindowToken(), 0);
                normalLogin();
                return true;
            }
            return false;
        });


        btnSign.setOnClickListener(new NoDoubleClickListener() {
            @Override
            protected void onNoDoubleClick(View v) {
                if (!NetworkUtil.isConnected()) {
                    ToastUtil.showToast("当前无网络");
                    return;
                }
                userAccount = etAccount.getText().toString();
                userPassword = etPassword.getText().toString();
                if (TextUtils.isEmpty(userAccount) || TextUtils.isEmpty(userPassword)) {
                    ToastUtil.showToast("账号或者密码不能为空");
                    return;
                }
                MyUser myUser = new MyUser();
                myUser.setUsername(userAccount);
                myUser.setPassword(userPassword);
                LoginRepository.INSTANCE.signByUser(myUser, new ILoginCallback<MyUser>() {
                    @Override
                    public void success(MyUser result) {
                        ToastUtil.showToast("sign success,welcome:" + result.getUsername());
                        EventBus.getDefault().post(new UpdateUserInfoEvent(true, result));
                    }

                    @Override
                    public void error(Throwable throwable) {
                        BmobException exception = (BmobException) throwable;
                        ToastUtil.showToast("sign fail," + exception.toString());
                    }
                });
            }

            @Override
            protected void onDoubleClick() {

            }
        });
        btnLogin.setOnClickListener(new NoDoubleClickListener() {
            @Override
            protected void onNoDoubleClick(View v) {
                normalLogin();
            }

            @Override
            protected void onDoubleClick() {

            }
        });
        loadingDialog = new LoadingDialog(this);
        loadingDialog.setLoadingText("请稍后")
                .setSuccessText("请求成功")
                .closeFailedAnim()
                .closeSuccessAnim()
                .setFailedText("请求失败");
    }

    private void normalLogin() {
        if (!NetworkUtil.isConnected()) {
            ToastUtil.showToast("当前无网络");
            return;
        }
        userAccount = etAccount.getText().toString();
        userPassword = etPassword.getText().toString();
        if (TextUtils.isEmpty(userAccount) || TextUtils.isEmpty(userPassword)) {
            ToastUtil.showToast("账号或者密码不能为空");
            return;
        }
        MyUser myUser = new MyUser();
        myUser.setUsername(userAccount);
        myUser.setPassword(userPassword);
        LoginRepository.INSTANCE.loginByUser(myUser, new ILoginCallback<MyUser>() {
            @Override
            public void success(MyUser result) {
                ToastUtil.showToast("login success,welcome:" + result.getUsername());
                EventBus.getDefault().post(new UpdateUserInfoEvent(true, result));
            }

            @Override
            public void error(Throwable throwable) {
                BmobException exception = (BmobException) throwable;
                ToastUtil.showToast("login fail," + exception.toString());
            }
        });
    }

    private void thirdLogin() {
        ivIcon = findViewById(R.id.iv_login_icon);
        ImageView ivWeibo = findViewById(R.id.iv_login_weibo);
        ivWeibo.setOnClickListener(new NoDoubleClickListener() {
            @Override
            protected void onNoDoubleClick(View v) {
                if (!NetworkUtil.isConnected()) {
                    ToastUtil.showToast("当前无网络");
                    return;
                }
                LoginRepository.INSTANCE.loginByShareSdk(SinaWeibo.NAME, iSdkLoginCallback);
                loadingDialog.show();
            }

            @Override
            protected void onDoubleClick() {

            }

        });
    }

    private void updateUserInfo(MyUser myUser) {
        Log.i(TAG, "third User: " + myUser.toString());
        LoginRepository.INSTANCE.updateUserByNewUser(myUser, new ILoginCallback<MyUser>() {
            @Override
            public void success(MyUser result) {
                loadingDialog.loadSuccess();
                EventBus.getDefault().post(new UpdateUserInfoEvent(true, result));
            }

            @Override
            public void error(Throwable throwable) {
                loadingDialog.loadFailed();
            }
        });
    }

    private void initCallback() {
        iSdkLoginCallback = new ILoginCallback<MyUser>() {
            @Override
            public void success(MyUser myUser) {
                CommonPref.instance().putString(LoginRepository.INSTANCE.getLAST_PLATFORM(), SinaWeibo.NAME);
                updateUserInfo(myUser);
            }

            @Override
            public void error(Throwable throwable) {
                loadingDialog.loadFailed();
            }
        };
    }


    @Subscribe(threadMode = ThreadMode.MAIN)
    public void LoginCallback(UpdateUserInfoEvent updateUserInfoEvent) {
        if (updateUserInfoEvent.loginResult) {
            GlideApp.with(this).load(updateUserInfoEvent.myUser.avatar).placeholder(R.drawable.icon_default_avatar).into(ivIcon);
            finish();
        }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        iSdkLoginCallback = null;
        loadingDialog.close();
    }
}

File: app/src/main/java/com/example/vivic/nolost/Login/LoginActivity.kt
package com.example.vivic.nolost.Login


import android.content.Context
import android.os.Bundle
import android.util.Log
import android.view.View
import android.view.inputmethod.EditorInfo
import android.view.inputmethod.InputMethodManager
import cn.bmob.v3.exception.BmobException
import cn.sharesdk.sina.weibo.SinaWeibo
import cn.sharesdk.tencent.qq.QQ
import com.example.vivic.nolost.GlideApp
import com.example.vivic.nolost.R
import com.example.vivic.nolost.activity.BaseActivity
import com.example.vivic.nolost.bean.MyUser
import com.example.vivic.nolost.commonUtil.BindEventBus
import com.example.vivic.nolost.commonUtil.NetworkUtil
import com.example.vivic.nolost.commonUtil.NoDoubleClickListener
import com.example.vivic.nolost.commonUtil.pref.CommonPref
import com.example.vivic.nolost.commonUtil.toastUtil.ToastUtil
import com.example.vivic.nolost.userCenter.UserRepository
import com.xiasuhuei321.loadingdialog.view.LoadingDialog
import io.reactivex.disposables.CompositeDisposable
import kotlinx.android.synthetic.main.activity_login.*
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import org.greenrobot.eventbus.ThreadMode

@BindEventBus
class LoginActivity : BaseActivity() {


    companion object {
        val TAG = LoginActivity::class.java.simpleName
    }

    private var loadingDialog: LoadingDialog? = null
    private var currentThirdPlatform: String? = null
    private val compositeDisposable: CompositeDisposable by lazy {
        CompositeDisposable()
    }
    private val inputMethodManager: InputMethodManager by lazy {
        getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_login)
        initview()
    }


    private fun initview() {
        cl_login_root.setOnClickListener { v -> inputMethodManager.hideSoftInputFromWindow(et_login_account!!.windowToken, 0) }
        et_login_password.setOnEditorActionListener { v, actionId, event ->
            if (actionId == EditorInfo.IME_ACTION_DONE) {
                inputMethodManager.hideSoftInputFromWindow(et_login_password!!.windowToken, 0)
                normalLogin()
            }
            false
        }

        btn_sign.setOnClickListener {
            if (!NetworkUtil.isConnected()) {
                ToastUtil.showToast("当前无网络")
                return@setOnClickListener
            }
            if (et_login_account.text.toString().isEmpty() || et_login_password.text.toString().isEmpty()) {
                ToastUtil.showToast("请输入完整信息")
                return@setOnClickListener
            }

            val myUser = MyUser().apply {
                this.username = et_login_account.text.toString()
                this.setPassword(et_login_password.text.toString())
            }
            compositeDisposable.add(UserRepository.signByUser(myUser, object : IBmobCallback<MyUser> {
                override fun success(result: MyUser?) {
                    ToastUtil.showToast("sign success,welcome:" + result?.username)
                    EventBus.getDefault().post(UserEvent(true, result))
                }

                override fun error(throwable: Throwable?) {
                    val exception = throwable as BmobException
                    ToastUtil.showToast("sign fail," + exception.toString())
                }
            }))
        }


        btn_login.setOnClickListener { normalLogin() }
        loadingDialog = LoadingDialog(this)
        loadingDialog!!.setLoadingText("请稍后")
                .setSuccessText("请求成功")
                .closeFailedAnim()
                .closeSuccessAnim()
                .setFailedText("请求失败")

        iv_login_weibo.setOnClickListener {
            if (!NetworkUtil.isConnected()) {
                ToastUtil.showToast("当前无网络")
                return@setOnClickListener
            }
            currentThirdPlatform = SinaWeibo.NAME
            UserRepository.loginByShareSdk(SinaWeibo.NAME, object : IBmobCallback<MyUser> {
                override fun success(result: MyUser?) {
                    updateUserInfo(result!!)
                }

                override fun error(throwable: Throwable?) {
                    loadingDialog?.loadFailed()
                }
            })
            loadingDialog?.show()
        }


        iv_login_qq.setOnClickListener(object : NoDoubleClickListener() {
            override fun onNoDoubleClick(v: View) {
                if (!NetworkUtil.isConnected()) {
                    ToastUtil.showToast("当前无网络")
                    return
                }
                currentThirdPlatform = QQ.NAME
                UserRepository.loginByShareSdk(QQ.NAME, object : IBmobCallback<MyUser> {
                    override fun success(result: MyUser?) {
                        updateUserInfo(result!!)
                    }

                    override fun error(throwable: Throwable?) {
                        loadingDialog?.loadFailed()
                    }
                })
                loadingDialog?.show()
            }

            override fun onDoubleClick() {

            }
        })
    }

    private fun normalLogin() {
        if (!NetworkUtil.isConnected()) {
            ToastUtil.showToast("当前无网络")
            return
        }
        if (et_login_account.text.toString().isEmpty() || et_login_password.text.toString().isEmpty()) {
            ToastUtil.showToast("请输入完整信息")
            return
        }
        val myUser = MyUser().apply {
            this.username = et_login_account.text.toString()
            this.setPassword(et_login_password.text.toString())
        }
        compositeDisposable.add(UserRepository.loginByUser(myUser, object : IBmobCallback<MyUser> {
            override fun success(result: MyUser?) {
                ToastUtil.showToast("login success,welcome:" + result?.username)
                EventBus.getDefault().post(UserEvent(true, result))
            }

            override fun error(throwable: Throwable?) {
                val exception = throwable as BmobException
                ToastUtil.showToast("login fail," + exception.toString())
            }
        }))
    }


    /**
     * 第三方登录成功之后，更改用户信息为第三方平台信息，并抛出通知
     *
     * @param myUser 用户信息为第三方平台信息
     */
    private fun updateUserInfo(myUser: MyUser) {
        Log.i(TAG, "third User: " + myUser.toString())
        compositeDisposable.add(UserRepository.updateUserByNewUser(myUser, object : IBmobCallback<MyUser> {
            override fun success(result: MyUser?) {
                loadingDialog?.loadSuccess()
                CommonPref.instance()?.putString(UserRepository.LAST_PLATFORM, currentThirdPlatform!!)
                EventBus.getDefault().post(UserEvent(true, result))
            }

            override fun error(throwable: Throwable?) {
                loadingDialog?.loadFailed()
            }
        }))
    }


    @Subscribe(threadMode = ThreadMode.MAIN)
    fun LoginCallback(userEvent: UserEvent) {
        if (userEvent.loginResult) {
            GlideApp.with(this).load(userEvent.myUser.avatar).placeholder(R.drawable.icon_default_avatar).into(iv_login_icon)
            finish()
        }
    }

    override fun onBackPressed() {
        super.onBackPressed()
        loadingDialog?.loadFailed()
        compositeDisposable.clear()
    }

    override fun onDestroy() {
        super.onDestroy()
        loadingDialog?.close()
        compositeDisposable.clear()
    }
}
